New tools - installation

iriark01 opened the pull request:
~Pending on Andy making a decision about the product name, but ready for engineering review.~
Mbed CLI 2

iriark01 committed:
Place holder file for first draft

iriark01 committed:
Install and upgrade

iriark01 committed:
Install & tools intro

iriark01 committed:
Install changes

Patater commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,53 @@
+
+# Install or upgrade
+
+Mbed Tools is Python based, so you can install it with pip.
```
"The Mbed Tools are Python based"?

iriark01 replied:
But it's a single tool. 
Which is one of the problems I had with the name: It's a plural name but there's just one tool. 
Hence waiting on Andy.

rwalton-arm replied:
"The new command line tool is a Python package named `mbed-tools`, so you can install it with pip."?

iriark01 replied:
We're gonna wait for a proper decision about name

Patater commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,53 @@
+- Pip (if not included in your Python installation). [Install for all operating systems](https://pip.pypa.io/en/stable/installing/).
+- cmake. [Install for all operating systems](https://cmake.org/install/).
+- Ninja [Install for all operating systems](https://github.com/ninja-build/ninja/wiki/Pre-built-Ninja-packages).
+<!--are there versions for cmake and ninja?-->
```
At least version 3.18.1 for CMake

v1.x for Ninja should suffice

Patater commented on docs/tools/tools_intro.md:
```diff
@@ -19,7 +19,18 @@ The Mbed Online Compiler is our in-house IDE and should be familiar to anyone wh
+Mbed OS has two versions of its command line tool:
+
+- For 6.y and newer, use the new build tools.
+- For 6.x and older, use Mbed CLI.
```
It's not a hard switch. We would support Mbed CLI for 6.y and newer as well. It's only with perhaps 7.x (although not decided yet) that we'd drop Mbed CLI support. We will recommend using the new tools over the old ones in a future Mbed OS release, but not yet.

iriark01 commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,53 @@
+
+- Python 3. Install for [Windows](https://docs.python.org/3/using/windows.html), [Linux](https://docs.python.org/3/using/unix.html) or [macOS](https://docs.python.org/3/using/mac.html).
+- Pip (if not included in your Python installation). [Install for all operating systems](https://pip.pypa.io/en/stable/installing/).
+- cmake. [Install for all operating systems](https://cmake.org/install/).
```
```suggestion
- cmake. [Install version 3.18.1 or newer for all operating systems](https://cmake.org/install/).
```

iriark01 commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,53 @@
+- Python 3. Install for [Windows](https://docs.python.org/3/using/windows.html), [Linux](https://docs.python.org/3/using/unix.html) or [macOS](https://docs.python.org/3/using/mac.html).
+- Pip (if not included in your Python installation). [Install for all operating systems](https://pip.pypa.io/en/stable/installing/).
+- cmake. [Install for all operating systems](https://cmake.org/install/).
+- Ninja [Install for all operating systems](https://github.com/ninja-build/ninja/wiki/Pre-built-Ninja-packages).
```
```suggestion
- Ninja [Install version 1.0 or newer for all operating systems](https://github.com/ninja-build/ninja/wiki/Pre-built-Ninja-packages).
```

iriark01 commented on docs/tools/tools_intro.md:
```diff
@@ -19,7 +19,18 @@ The Mbed Online Compiler is our in-house IDE and should be familiar to anyone wh
+
+Mbed OS has two versions of its command line tool:
+
+- For 6.y and newer, use the new build tools.
```
```suggestion
- For 6.y and newer, use the new build tools or Mbed CLI.
```

rwalton-arm commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,53 @@
+
+<!--well this is a disaster-->
+
+Starting with version 6.x, Mbed OS is moving from Mbed CLI to a new build tool: it's cmake-based,
```
At the moment I'm finding the way this is worded slightly confusing, we say our new "build tool" is cmake-based, then below we say it's Python based. We should make the distinction between CMake and the Python command line tools clearer.

iriark01 replied:
The whole intro is a bit of a mess (hence the "well this is a disaster"). When I get more info on the tool, it'll be easier to know what to say here. But if you can explain the distinction, that'll help.

rwalton-arm replied:
Sorry, I should've added an explanation. At a high level the distinction is:
* Ninja is our build system of choice.
* CMake is an open-source system that generates the build environment and manages the build process in a compiler independent manner.
* mbed-tools (or whatever we end up calling it) is our new command line tool that parses the build configuration from the Mbed configuration system and outputs it in a format CMake can read. mbed-tools will also provide a user friendly interface to CMake and Ninja, allowing the user to configure, generate and execute the build system with a single command.

Hopefully that helps!

iriark01 replied:
That will be lovely for the intro page (the next file in this review), and should help me figure out what people are installing here. Thank you!

rwalton-arm commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,53 @@
+Mbed Tools is Python based, so you can install it with pip.
+<!--other than being a super vague name, it's a problem that it's a single tool with a plural name. "Mbed Tools is"  just parses as an error!-->
+
+<span class="tips">**Tip:** We recommend using a virtual environment such as [Pipenv](https://github.com/pypa/pipenv/blob/master/README.md) to avoid Python dependency conflicts.</span>
```
`pipenv` isn't a virtual environment, it's a virtual environment management tool.

iriark01 replied:
Do we have a tool we want to recommend, then?

rwalton-arm replied:
We do want to recommend `pipenv` to manage the development environment, at the moment, but we don't want to refer to `pipenv` as "a virtual environment". A virtual environment is a self-contained directory tree that contains a Python installation for a particular version of Python, plus a number of additional packages. `pipenv` is a "virtual environment management tool". We probably want to say something along the lines of: "We recommend using a virtual environment to avoid any Python dependency conflicts. We recommend the `pipenv` tool to manage your virtual development environment."

iriark01 committed:
Update docs/tools/new_tools_temp.md

iriark01 committed:
Update docs/tools/new_tools_temp.md

iriark01 committed:
Update docs/tools/tools_intro.md

iriark01 committed:
Adding Rob's excellent explanations

iriark01 commented:
@rwalton-arm please review again

rwalton-arm reviewed (commented):
LGTM

loverdeg commented:
IMO the 'tools' being Mbed specific is part of the root cause of this strange issue.

iriark01 committed:
Questions about using the tool

jainvikas8 commented on docs/tools/new_tools_temp.md:
````diff
@@ -0,0 +1,144 @@
+    mbed-tools init -c <PATH>
+    ```
+
+    <!--so how do I tell it where the existing copy is? In Mbed Studio you have to explicitly "tell it" where the local copy is-->
````
Example: 
If we have a mbed-os-example-blinky application and we may execute it as `mbed-tools init .` inside the folder.

rwalton-arm replied:
The tool doesn't support "telling it where an existing copy is" yet, it (meaning an existing copy of mbed-os) would have to be symlinked into the program folder. We are going to support this at some point when we implement "workspace" support, but it's not done yet.

iriark01 replied:
How actually do I get an example going? How do I tell it to go download a copy of mbed-os-example-blinky, and do I need to use `init` in an example?

iriark01 replied:
How do we use SymLink, then? If it's not been done before in Mbed OS, we need to explain it.

rwalton-arm replied:
`init` is just for creating a new project, it's similar to `mbed new`. We also have a `clone` command for downloading examples, which is similar to `mbed import` but we haven't documented that command, yet. I think we should just say "copy" rather than symbolic linking as I'm not sure we've tested symlinked CMake builds work yet. The `-c` workflow with the `init` command won't be supported until we have workspace support in the tools, which isn't implemented yet.

jainvikas8 commented on docs/tools/new_tools_temp.md:
````diff
@@ -0,0 +1,144 @@
+mbed-tools devices
+```
+
+<!--why? what do I do with this info? is this for the config?-->
````
This is the for `mbed-tools config` or to list connected devices.  (info)

iriark01 replied:
What does the output look like? Does it give me the target name so I can use it for my config command?

jainvikas8 replied:
Yes, target name is given.

jainvikas8 commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,144 @@
+
+## Set environment variables
+
+<!--like what? when? why? should this be with installation instructions?-->
```
The variables used by Mbed Tools can be set either directly via environment variables or
using a `.env` file containing the variable definitions as follows:

VARIABLE=value

Environment variables take precedence, meaning the values set in the file will be overridden
by any values previously set in your environment.

There's also a way to access "private" boards listed in the online DB. `mbed-tools env`, prints an explanation of how to do that using an authorisation token.

iriark01 replied:
I think this needs to be in the setup stage; you wouldn't expect people to constantly set environment variables, would you? It certainly doesn't strike me as a step in every single project.

jainvikas8 replied:
yes, it doesn't need to be set always...

jainvikas8 commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,144 @@
+
+## Configure Mbed OS
+
+<!--I don't understand what this actually does. Does it generate a config file? why do I need to do this? what does "for use with" mean - do I use Mbed Tools with the target and toolchain, or am I configuring Mbed OS itself? Do I always have to do this before I build, or can I do it at the same time as the build, or can I just skip it sometimes?-->
```
Does it generate a config file?  -- yes
what does "for use with" mean -- It means we use this with specific target and toolchain
am I configuring Mbed OS itself?  -- no, this is configuring build system for a particular target with a toolchain before building it.
Do I always have to do this before I build? -- No, once configured you don't need to, unless you need to change the parameters - target,, toolchain.... etc...

jainvikas8 commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,144 @@
+
+Currently the supported targets are `K64F`, `DISCO_L475VG_IOT01A`, `NRF52840_DK`
+
+<!--just for configuring, or can you only build with those?-->
```
Config and build.

iriark01 replied:
Okay, we need to mention this much earlier. I'll add a line.

jainvikas8 commented on docs/tools/new_tools_temp.md:
````diff
@@ -0,0 +1,144 @@
+cmake -S . -B cmake_build -GNinja -DCMAKE_BUILD_TYPE=<profile>
+cmake --build cmake_build
+```
+<!--what are all the other parameters?-->
````
These are Cmake options
```   
Options
       -S <path-to-source>
              Path to root directory of the CMake project to build.
       -B <path-to-build>
              Path to directory which CMake will use as the root of build
              directory.
              If the directory doesn't already exist CMake will make it.
```

-GNinja -- we are using Ninja tool
-DCMAKE_BUILD_TYPE -- Build type

jainvikas8 commented on docs/tools/new_tools_temp.md:
````diff
@@ -0,0 +1,144 @@
+cmake --build cmake_build
+```
+<!--what are all the other parameters?-->
+<!--what does the first command do, and what does the second one do?-->
````
`cmake -S . -B cmake_build -GNinja -DCMAKE_BUILD_TYPE=<profile>`
This one sets the build parameters

the second does the actual build.

iriark01 replied:
Okay. I will show these as two steps, then, with an explanation.

rwalton-arm commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,144 @@
+
+## Configure Mbed OS
+
+<!--I don't understand what this actually does. Does it generate a config file? why do I need to do this? what does "for use with" mean - do I use Mbed Tools with the target and toolchain, or am I configuring Mbed OS itself? Do I always have to do this before I build, or can I do it at the same time as the build, or can I just skip it sometimes?-->
```
It parses the mbed config system (mbed_lib.json, mbed_app.json and targets.json) for a particular target and toolchain, and outputs a cmake script that is used to configure the build system. After the cmake script has been generated you can reuse it if you're building for the same target and toolchain. If you decide to change target and/or toolchain, you have to run the configure command again. Note: we will soon be implementing a build command to simplify the config and build steps so the user doesn't have to run multiple commands. The build command isn't implemented yet.

iriark01 replied:
```suggestion
The Mbed OS configuration system parses the configuration files in your project (mbed_lib.json, mbed_app.json and targets.json) for a particular target and toolchain, and outputs a CMake script. The build system uses this script to build for your target, using your toolchain. 

If you're rebuilding for the same target and toolchain, you can keep using the same CMake script, so you won't have to use the `configure` command again for each build. If you change your target or toolchain, run the `configure` command again to generate a new CMake script.```

rwalton-arm commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,144 @@
+
+## Build using CMake
+
+After preparing the Mbed configuration information, you can use CMake to build your application:<!--can, but don't have to? what are the other options?-->
```
There are no other options, you *must* use CMake to generate the build environment and invoke the build system (which is Ninja in our case); the configuration script the tool generates with the `configure` command is CMake specific.

iriark01 replied:
```suggestion
After preparing the Mbed configuration information, use CMake to build your application:
```

rwalton-arm commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,144 @@
+<!--what are all the other parameters?-->
+<!--what does the first command do, and what does the second one do?-->
+- We recommend using a dedicated **cmake_build** folder to keep things organised.<!--so do I replace cmake_build with a new value?-->
+- The `profile` value can be `release`, `debug` or `develop`.<!--same as normal, or something specific to CMake? https://os.mbed.com/docs/mbed-os/v6.2/program-setup/build-profiles-and-rules.html-->
```
Same as the old profiles.

rwalton-arm commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,144 @@
+- We recommend using a dedicated **cmake_build** folder to keep things organised.<!--so do I replace cmake_build with a new value?-->
+- The `profile` value can be `release`, `debug` or `develop`.<!--same as normal, or something specific to CMake? https://os.mbed.com/docs/mbed-os/v6.2/program-setup/build-profiles-and-rules.html-->
+
+<!--what does this generate and where? is this where we add instructions for flashing the board?-->
```
It generates the binary and hex files in the `cmake_build` folder. "Flashing" (i.e invoking `cp` to copy the binary to the USB disk presented by DAPlink) isn't implemented in the new tools yet, but it's on our backlog.

iriark01 replied:
So flashing is drag&drop?

rwalton-arm replied:
At the moment flashing is not supported by the tool and is left to the user. They can use drag&drop.

rwalton-arm commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,144 @@
+
+## Example applications
+
+<!--can I create a new one or do I have to use an example? I'm asking because these were listed as part of the build instructions file, I think-->
```
You can create a new project and build it, that's what the `init` command does. This is just a list of the examples that currently support CMake.

iriark01 replied:
How do I actually get them? What's the new `mbed import`?

rwalton-arm replied:
We have a `clone` command which is the new `mbed import`. I don't think we have documented that command yet.

rwalton-arm commented on docs/tools/tools_intro.md:
```diff
@@ -19,7 +19,25 @@ The Mbed Online Compiler is our in-house IDE and should be familiar to anyone wh
+
+Starting with version 6.x, Mbed OS is moving from Mbed CLI to a new build tool: <!--that has no proper name yet-->. It uses Ninja as a build system, and CMake to generate the build environment and manage the build process in a compiler-independent manner. <!--need better phrasing-->
+
+Mbed Tools parses the Mbed OS build configuration and outputs it to a format CMake can read. It also provides a use friendly interface to CMake and Ninja so you can configure, generte and execute builds with a single command.
```
```suggestion
Mbed Tools parses the Mbed OS build configuration and outputs it to a format CMake can read. It also provides a user friendly interface to CMake and Ninja so you can configure, generate and execute builds with a single command.
```

iriark01 commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,144 @@
+
+<!--I don't understand what this actually does. Does it generate a config file? why do I need to do this? what does "for use with" mean - do I use Mbed Tools with the target and toolchain, or am I configuring Mbed OS itself? Do I always have to do this before I build, or can I do it at the same time as the build, or can I just skip it sometimes?-->
+
+To prepare the Mbed configuration information for use with a specific target and toolchain:
```
```suggestion
To generate a script for a new target and toolchain combination:
```

jainvikas8 replied:
script = Application configuration ?

iriark01 commented on docs/tools/new_tools_temp.md:
````diff
@@ -0,0 +1,144 @@
+python -m pip install mbed-tools --upgrade
+```
+
+# Use
````
```suggestion
# Use

Currently, the new Mbed Tools supports three boards: `K64F`, `DISCO_L475VG_IOT01A` and `NRF52840_DK`.

```

jainvikas8 replied:
'these' boards would be good, rather than being specific?

iriark01 commented on docs/tools/new_tools_temp.md:
````diff
@@ -0,0 +1,144 @@
+To prepare the Mbed configuration information for use with a specific target and toolchain:
+
+```
+mbed-tools configure -m <target> -t <toolchain>
````
What are the possible toolchain values? Same as for Mbed CLI?

jainvikas8 replied:
Yes, GCC_ARM and ARM

iriark01 commented on docs/tools/new_tools_temp.md:
````diff
@@ -0,0 +1,144 @@
+```
+<!--what are all the other parameters?-->
+<!--what does the first command do, and what does the second one do?-->
+- We recommend using a dedicated **cmake_build** folder to keep things organised.<!--so do I replace cmake_build with a new value?-->
````
This question was missed: do I replace `cmake_build` with a new value?

rwalton-arm replied:
Sorry, not sure I understand the question. We create the `cmake_build` folder with the `cmake -B cmake_build` command. Why would we need to replace it with a new value? I assume this recommendation is here so people don't do an "in source" build and pollute the source tree with build system artifacts.

jainvikas8 replied:
`cmake_build` is the folder. It is recommended to use a folder for the build, as this will keep things organised. Normally a default name of that folder is `cmake_build`. 
We mentioned in the command `cmake --build cmake_build` because the user can copy and paste to execute it.

iriark01 committed:
New info and new questions

jainvikas8 commented on docs/tools/new_tools_temp.md:
````diff
@@ -0,0 +1,186 @@
+    mbed-tools init -c <PATH>
+    ```
+
+<!--this seems to be an absolute path - had to start with `~/`. Can they work with relative paths? can they work with "here" to tell mbed-tools to work in the current directory?-->
````
`mbed-tools init -c .` can be used and also with relative paths like `mbed-tools init -c .\mbed-os-example-blinky`

jainvikas8 commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,186 @@
+
+    This generates two files: BIN and HEX in the build output directory (`cmake_build` in this example).
+
+1. Drag and drop the generated file to your board. <!--which file? is it up to them?-->
```
Mostly it's in the format of **application.hex** for Example:  `mbed-os-example-blinky.hex`

iriark01 replied:
What does "mostly" mean? Does it depend on the board type?

jainvikas8 replied:
Sorry for the confusion, but it is in **application.hex** format.

rwalton-arm commented on docs/tools/new_tools_temp.md:
````diff
@@ -0,0 +1,186 @@
+    mbed-tools init -c <PATH>
+    ```
+
+<!--this seems to be an absolute path - had to start with `~/`. Can they work with relative paths? can they work with "here" to tell mbed-tools to work in the current directory?-->
````
It works with relative paths... You have to at least specify the directory name to create as it uses that as the project name. `mbedtools init -c test` works.

iriark01 replied:
I'm not sure you and Vikas gave exactly the same instructions. So the options are:
relative:
`.` for "here"
`.\folder` for "new folder here"
or absolute

rwalton-arm commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,186 @@
+<!--this seems to be an absolute path - had to start with `~/`. Can they work with relative paths? can they work with "here" to tell mbed-tools to work in the current directory?-->
+
+
+<!--If I can't symlink - if I have to copy my local version of Mbed OS - then this is less about savings space and more about allowing me to take a non-default version of Mbed OS, then?-->
```
It will be about saving space, but we haven't implemented workspace support yet..

iriark01 replied:
Might be best to remove these instructions, then.

rwalton-arm commented on docs/tools/new_tools_temp.md:
````diff
@@ -46,23 +46,49 @@ Use pip to install:
+To override default environment variables:
+
+```
+mbed-tools env <VARIABLE>=<value>
````
What's this? `mbedtools env` just prints information about the environment variables you can set if you want to access private boards in the online database, it doesn't allow you to set variables like that...

iriark01 replied:
So. How and why do you actually use this idea? Because the impression I got from the original text is that env is how you modify things, not list them.

rwalton-arm replied:
You can either specify the variables in a `.env` file and the tool will discover them, or you can set them per terminal session the same way you would set any environment variable https://www.schrodinger.com/kb/1842. The "why" is what `mbedtools env` tells you, as pasted in the comment below. `mbedtools env` doesn't list any set variables, it only describes the variables the tool recognises and why you might need to set them (most users will not need to set them at all).

iriark01 replied:
So the command is `setenv`, rather than `env`? (that's according to the KB you linked to)

rwalton-arm commented on docs/tools/new_tools_temp.md:
```diff
@@ -46,23 +46,49 @@ Use pip to install:
 
+## Set environment variables
+
+<!--like what? when? why? should this be with installation instructions?-->
```
If you run `mbedtools env` it will print a detailed description to the console...

```
$ mbedtools env
These variables used by Mbed Tools can be set either directly via environment variables or
using a `.env` file containing the variable definitions as follows:

VARIABLE=value

Environment variables take precedence, meaning the values set in the file will be overriden
by any values previously set in your environment.

+---------------------------------------------------------------------------------+
| Do not upload `.env` files containing private tokens to version control! If you |
| use this package as a dependency of your project, please ensure to include the  |
| `.env` in your `.gitignore`.                                                    |
+---------------------------------------------------------------------------------+


MBED_API_AUTH_TOKEN

        Token to use when accessing online API.

        Mbed Targets uses the online mbed board database at os.mbed.com as its data source.
        A snapshot of the board database is shipped with the package, for faster lookup of known
        boards. Only public boards are stored in the database snapshot. If you are fetching data
        for a private board, mbed-targets will need to contact the online database.

        To fetch data about private boards from the online database, the user must have an account
        on os.mbed.com and be member of a vendor team that has permissions to see the private board.
        An authentication token for the team member must be provided in an environment variable named
        `MBED_API_AUTH_TOKEN`.

MBED_DATABASE_MODE

        Database mode to use when retrieving board data.

        Mbed Targets supports an online and offline mode, which controls where to look up the board database.

        The board lookup can be from either the online or offline database, depending
        on the value of an environment variable called `MBED_DATABASE_MODE`.

        The mode can be set to one of the following:

        - `AUTO`: the offline database is searched first, if the board isn't found the online database is searched.
        - `ONLINE`: the online database is always used.
        - `OFFLINE`: the offline database is always used.

        If `MBED_DATABASE_MODE` is not set, it defaults to `AUTO`.
```

rwalton-arm commented on docs/tools/new_tools_temp.md:
```diff
@@ -78,67 +104,83 @@ To create a new Mbed OS project in a specified path:
+
+### Use an example application
+
+To create a local copy of an example application, use the `clone` command with the full GitHub URL listed below:
```
You don't need to give the full github URL, you can use `mbedtools clone <example-name>` for ARMmbed examples.

rwalton-arm commented on docs/tools/new_tools_temp.md:
```diff
@@ -46,23 +46,49 @@ Use pip to install:
     python -m pip install mbed-tools --pre
```
Do we mention anywhere that `mbed-tools` is only compatible with python 3.6 and above?

iriark01 replied:
We say Python 3, not 3.6. Line 21

rwalton-arm replied:
OK. We should specify Python 3.6 or later, as it won't work correctly with Python 3.5 or below. Annoyingly Python 3 has many changes to standard library APIs between 3.0 and 3.6 that aren't entirely backwards compatible.

jainvikas8 commented on docs/tools/new_tools_temp.md:
`````diff
@@ -0,0 +1,186 @@
+````
+
+- Blinky: [https://github.com/ARMmbed/mbed-os-example-blinky](https://github.com/ARMmbed/mbed-os-example-blinky)
+- BLE button: [https://github.com/ARMmbed/mbed-os-example-ble/tree/master/BLE_Button](https://github.com/ARMmbed/mbed-os-example-ble/tree/master/BLE_Button)
`````
This needs to be changed to 
```
- BLE button: [mbed-os-example-ble](https://github.com/ARMmbed/mbed-os-example-ble)
```
Cloning with `https://github.com/ARMmbed/mbed-os-example-ble/tree/master/BLE_Button` will not work.

Also only BLE button example is only supported.

iriark01 replied:
```suggestion
- [mbed-os-example-ble](https://github.com/ARMmbed/mbed-os-example-ble) - use the BLE button example.
```

iriark01 commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,186 @@
+
+## Prerequisite
+
+- Python 3. Install for [Windows](https://docs.python.org/3/using/windows.html), [Linux](https://docs.python.org/3/using/unix.html) or [macOS](https://docs.python.org/3/using/mac.html).
```
```suggestion
- Python 3.6 or newer. Install for [Windows](https://docs.python.org/3/using/windows.html), [Linux](https://docs.python.org/3/using/unix.html) or [macOS](https://docs.python.org/3/using/mac.html).
```

iriark01 commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,186 @@
+
+## Set environment variables
+
+<!--like what? when? why? should this be with installation instructions?-->
```
Mbed Tools has two environment variables: 

- `MBED_API_AUTH_TOKEN`: Token to access private board information stored for a vendor team. <!--where do they get this token?-->
-  `MBED_DATABASE_MODE`: Use online or offline mode. Possible values:
        - `AUTO`: Search the offline database first; search the online database only if the board wasn't found offline. This is the default value.
        - `ONLINE`: Alway use the online database.
        - `OFFLINE`: Always use the offline database.

jainvikas8 replied:
The description is printed when `mbed-tools env` is used... I don't see why this information is required to be repeated.

iriark01 commented on docs/tools/new_tools_temp.md:
````diff
@@ -0,0 +1,186 @@
+To override default environment variables:
+
+```
+mbed-tools env <VARIABLE>=<value>
````
```suggestion
To list the current values:

`mbed-tools env `

To set values, create an `.env` file in the root directory of the project. The file should contain definitions in the `<VARIABLE>=<value>` format. 
```

iriark01 commented on docs/tools/new_tools_temp.md:
````diff
@@ -0,0 +1,186 @@
+mbed-tools env <VARIABLE>=<value>
+```
+
+You can also use a `.env` file containing definitions in the same `<VARIABLE>=<value>` format. Note that environment variables set using the command line will override those listed in the file.
````
```suggestion
```

iriark01 commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,186 @@
+
+### Use an example application
+
+To create a local copy of an example application, use the `clone` command with the full GitHub URL listed below:
```
```suggestion
To create a local copy of an example application, use the `clone` command with the example name listed below:
```

iriark01 commented on docs/tools/new_tools_temp.md:
````diff
@@ -0,0 +1,186 @@
+To create a local copy of an example application, use the `clone` command with the full GitHub URL listed below:
+
+```
+mbed-tools clone <link to GitHub example> <PATH>
````
```suggestion
mbed-tools clone <example> <PATH>
```

iriark01 commented on docs/tools/new_tools_temp.md:
`````diff
@@ -0,0 +1,186 @@
+mbed-tools clone <link to GitHub example> <PATH>
+````
+
+- Blinky: [https://github.com/ARMmbed/mbed-os-example-blinky](https://github.com/ARMmbed/mbed-os-example-blinky)
`````
```suggestion
- [mbed-os-example-blinky](https://github.com/ARMmbed/mbed-os-example-blinky)
```

iriark01 commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,186 @@
+
+- Blinky: [https://github.com/ARMmbed/mbed-os-example-blinky](https://github.com/ARMmbed/mbed-os-example-blinky)
+- BLE button: [https://github.com/ARMmbed/mbed-os-example-ble/tree/master/BLE_Button](https://github.com/ARMmbed/mbed-os-example-ble/tree/master/BLE_Button)
+- Cellular: [https://github.com/ARMmbed/mbed-os-example-cellular](https://github.com/ARMmbed/mbed-os-example-cellular)
```
```suggestion
- [mbed-os-example-cellular](https://github.com/ARMmbed/mbed-os-example-cellular)
```

iriark01 commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,186 @@
+- Blinky: [https://github.com/ARMmbed/mbed-os-example-blinky](https://github.com/ARMmbed/mbed-os-example-blinky)
+- BLE button: [https://github.com/ARMmbed/mbed-os-example-ble/tree/master/BLE_Button](https://github.com/ARMmbed/mbed-os-example-ble/tree/master/BLE_Button)
+- Cellular: [https://github.com/ARMmbed/mbed-os-example-cellular](https://github.com/ARMmbed/mbed-os-example-cellular)
+- DeviceKey: [https://github.com/ARMmbed/mbed-os-example-devicekey](https://github.com/ARMmbed/mbed-os-example-devicekey)
```
```suggestion
- [mbed-os-example-devicekey](https://github.com/ARMmbed/mbed-os-example-devicekey)
```

iriark01 commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,186 @@
+- BLE button: [https://github.com/ARMmbed/mbed-os-example-ble/tree/master/BLE_Button](https://github.com/ARMmbed/mbed-os-example-ble/tree/master/BLE_Button)
+- Cellular: [https://github.com/ARMmbed/mbed-os-example-cellular](https://github.com/ARMmbed/mbed-os-example-cellular)
+- DeviceKey: [https://github.com/ARMmbed/mbed-os-example-devicekey](https://github.com/ARMmbed/mbed-os-example-devicekey)
+- KVStore: [https://github.com/ARMmbed/mbed-os-example-kvstore](https://github.com/ARMmbed/mbed-os-example-kvstore)
```
```suggestion
- [mbed-os-example-kvstore](https://github.com/ARMmbed/mbed-os-example-kvstore)
```

iriark01 commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,186 @@
+- Cellular: [https://github.com/ARMmbed/mbed-os-example-cellular](https://github.com/ARMmbed/mbed-os-example-cellular)
+- DeviceKey: [https://github.com/ARMmbed/mbed-os-example-devicekey](https://github.com/ARMmbed/mbed-os-example-devicekey)
+- KVStore: [https://github.com/ARMmbed/mbed-os-example-kvstore](https://github.com/ARMmbed/mbed-os-example-kvstore)
+- LoRaWAN: [https://github.com/ARMmbed/mbed-os-example-lorawan](https://github.com/ARMmbed/mbed-os-example-lorawan)
```
```suggestion
- [mbed-os-example-lorawan](https://github.com/ARMmbed/mbed-os-example-lorawan)
```

iriark01 commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,186 @@
+- DeviceKey: [https://github.com/ARMmbed/mbed-os-example-devicekey](https://github.com/ARMmbed/mbed-os-example-devicekey)
+- KVStore: [https://github.com/ARMmbed/mbed-os-example-kvstore](https://github.com/ARMmbed/mbed-os-example-kvstore)
+- LoRaWAN: [https://github.com/ARMmbed/mbed-os-example-lorawan](https://github.com/ARMmbed/mbed-os-example-lorawan)
+- Mbed Crypto: [https://github.com/ARMmbed/mbed-os-example-mbed-crypto](https://github.com/ARMmbed/mbed-os-example-mbed-crypto)
```
```suggestion
- [mbed-os-example-mbed-crypto](https://github.com/ARMmbed/mbed-os-example-mbed-crypto)
```

iriark01 commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,186 @@
+- KVStore: [https://github.com/ARMmbed/mbed-os-example-kvstore](https://github.com/ARMmbed/mbed-os-example-kvstore)
+- LoRaWAN: [https://github.com/ARMmbed/mbed-os-example-lorawan](https://github.com/ARMmbed/mbed-os-example-lorawan)
+- Mbed Crypto: [https://github.com/ARMmbed/mbed-os-example-mbed-crypto](https://github.com/ARMmbed/mbed-os-example-mbed-crypto)
+- NFC: [https://github.com/ARMmbed/mbed-os-example-nfc](https://github.com/ARMmbed/mbed-os-example-nfc)
```
```suggestion
- [mbed-os-example-nfc](https://github.com/ARMmbed/mbed-os-example-nfc)
```

iriark01 commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,186 @@
+- LoRaWAN: [https://github.com/ARMmbed/mbed-os-example-lorawan](https://github.com/ARMmbed/mbed-os-example-lorawan)
+- Mbed Crypto: [https://github.com/ARMmbed/mbed-os-example-mbed-crypto](https://github.com/ARMmbed/mbed-os-example-mbed-crypto)
+- NFC: [https://github.com/ARMmbed/mbed-os-example-nfc](https://github.com/ARMmbed/mbed-os-example-nfc)
+- Sockets: [https://github.com/ARMmbed/mbed-os-example-sockets](https://github.com/ARMmbed/mbed-os-example-sockets)
```
```suggestion
- [mbed-os-example-sockets](https://github.com/ARMmbed/mbed-os-example-sockets)
```

iriark01 commented on docs/tools/new_tools_temp.md:
````diff
@@ -0,0 +1,186 @@
+    cmake --build cmake_build
+    ```
+
+    This generates two files: BIN and HEX in the build output directory (`cmake_build` in this example).
````
```suggestion
    This generates two files in the build output directory (`cmake_build` in this example): HEX and BIN. <!--find a place where we explain where HEX is useful and link to it-->.
```

jainvikas8 replied:
It generates both `HEX` and `bin` files, which can be used.

iriark01 replied:
But when I asked how you know which one to use, the answer I got was "it's HEX".

jainvikas8 replied:
The build will generate both hex and bin. It’s up to the developer to use either of them. But the recommendation is the `HEX` file.

iriark01 replied:
What's the advantage of the HEX file?

jainvikas8 replied:
The advantage with hex is specifying the address for the image. During flashing only the address range mentioned in the `hex` are modified.

iriark01 replied:
That actually sounds like a pretty big deal that we should maybe consider focusing on. Maybe not right in the middle of the build process, though. Maybe the build tools intro? Do they all do this?

iriark01 commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,186 @@
+
+    This generates two files: BIN and HEX in the build output directory (`cmake_build` in this example).
+
+1. Drag and drop the generated file to your board. <!--which file? is it up to them?-->
```
```suggestion
1. Drag and drop the generated file to your board. 
```

iriark01 committed:
Update docs/tools/new_tools_temp.md

iriark01 committed:
Update docs/tools/new_tools_temp.md

iriark01 committed:
Update docs/tools/new_tools_temp.md

iriark01 committed:
Apply suggestions from code review

iriark01 committed:
Update docs/tools/new_tools_temp.md

iriark01 commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,186 @@
+
+<!--like what? when? why? should this be with installation instructions?-->
+
+To override default environment variables:
```
```suggestion
Mbed Tools has two environment variables:

- `MBED_API_AUTH_TOKEN`: Token to access private board information stored for a vendor team.
- `MBED_DATABASE_MODE`: Use online or offline mode. Possible values:
    - AUTO: Search the offline database first; search the online database only if the board wasn't found offline. This is the default value.
    - ONLINE: Alway use the online database.
    - OFFLINE: Always use the offline database.```

iriark01 committed:
Apply suggestions from code review

iriark01 committed:
For Dev's review

iriark01 committed:
Env is the same as env --help, so removing duplication

urutva commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,189 @@
+    If you have already created a project folder, you can use `.`
+
+    If you want the `init` command to create a project folder, use `.\<folder-name>`.
+
```
@iriark01 We should also document `-c` option which allows sym-linking existing mbd-os. Please check https://github.com/ARMmbed/mbed-tools/issues/65

iriark01 replied:
I removed that because it wasn't fully functional

urutva replied:
@iriark01 Did you face any errors when you ran`mbed-tools init -c`?

iriark01 replied:
I was simply told it doesn't work yet

jainvikas8 replied:
This has been added recently, please review and resolve this comment @urutva

urutva replied:
@iriark01 @jainvikas8 I've tried `mbed-tools init -c` after commenting and it works as expected. It should be added to the document.

iriark01 replied:
It's in line 42 in the file about using it

urutva commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,189 @@
+
+The Mbed OS configuration system parses the configuration files in your project (mbed_lib.json, mbed_app.json and targets.json) for a particular target and toolchain, and outputs a CMake script. The build system uses this script to build for your target, using your toolchain.
+
+<span class="tips">**Tip:** If you're rebuilding for the same target and toolchain, you can keep using the same CMake script, so you won't have to use the `configure` command again for each build. If you change your target or toolchain, run the `configure` command again to generate a new CMake script.</span>
```
```suggestion
<span class="tips">**Tip:** If you're rebuilding for the same target and toolchain, you can keep using the same CMake script, so you won't have to use the `configure` command again for each build. If you change any of mbed_lib.json, mbed_app.json, targets.json, target or toolchain, run the `configure` command again to generate a new CMake script.</span>
```

jainvikas8 commented on docs/tools/new_tools_temp.md:
```diff
@@ -0,0 +1,189 @@
+
+To set values, create an `.env` file in the root directory of the project. The file should contain definitions in the `<VARIABLE>=<value>` format.
+
+For more information, use `mbed-tools env --help`
```
`mbed-tools env`

iriark01 replied:
Have we figured out what does what? I found the Slack conversation really confusing

urutva commented on docs/tools/new_tools_temp.md:
````diff
@@ -0,0 +1,189 @@
+```
+
+# Use
+
````
@iriark01 We should document the verbose logging somewhere. I'm not sure about the location. I think `Use` chapter might be a good fit.
Example:
`mbed-tools -v configure -m <target> -t <toolchain>`
TIP: repeat character `v` to increase logging level. For example `-vv` or `-vvv`.

iriark01 replied:
How many Vs can you use?

urutva replied:
It can be any number. Following might help:
```
None -> Show only errros
-v -> Show warnings + errors
-vv -> Show information + warnings + errors
-vvv -> Show debug + information + warnings + errors
```
Repeating more than thrice will have the same effect as repeating thrice. But `mbed-tools` won't throw any error for repeating more then thrice.

iriark01 committed:
Update docs/tools/new_tools_temp.md

Co-authored-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

iriark01 committed:
Verbose

iriark01 committed:
Verbose + comments

jainvikas8 commented on docs/tools/new_tools_temp.md:
```diff
@@ -187,3 +190,18 @@ Use CMake to build your application:
+
+## Logging
+
+To specify the log level, use the verbose logging option (`-v`) before the first argument:
```
```suggestion
To specify the log level, use the verbose logging option (`-v`) before the first argument.

If you don't use `-v`, by default the log will show only errors. These are the following log levels:
```

jainvikas8 replied:
```suggestion
To specify the log level, use the verbose logging option (`-v`) before the first argument.

If you don't use `-v`, by default the log will show only errors. These are the following log levels:
```

iriark01 committed:
HEX or BIN

jainvikas8 committed:
docs: BLE Button example is not supported

From Mbed OS 6.4.0 onwards BL Button example is not supported anymore,
therefore recommend to use BLE LED example instead

jainvikas8 committed:
doc: `env` command is been removed

The command `env` only provided information rather than any usability,
therefore remove it and explain in clear terms.

jainvikas8 committed:
docs: Add `build` cmd

Mbed OS can be configured and built using a single command

This improves usability.

jainvikas8 committed:
docs: add `libs` command

`libs` commands helps to list libraries dependencies with their URL and
git references.

jainvikas8 committed:
docs: `init` command update

This command can be used to create a project without downloading
a copy of Mbed OS (reuse an existing copy)

iriark01 committed:
New engineering content

iriark01 committed:
Starting to tidy up

iriark01 committed:
Missed one

iriark01 committed:
Some general pages

iriark01 committed:
JSON was wrong

iriark01 committed:
Links and redirects

iriark01 committed:
Links need .html

iriark01 committed:
Not dealing well with links this morning

jainvikas8 committed:
docs: Advise on virtual environment usage

There many virtual environment than pipenv such as Tox.

Rather be generic about this in the docs to the developer

jainvikas8 committed:
docs: Python development package

Installation of header files and static libraries are required
by Pip and done by python development package.

jainvikas8 committed:
change to python3 dev

jainvikas8 committed:
change wording

jainvikas8 committed:
docs: Add tip

Tip to make executable accessible after the installation.

iriark01 commented on docs/tools/mbed_cli_2/install.md:
```diff
@@ -0,0 +1,53 @@
+
+## Install
+
+Some header files and static libraries are required by Pip and is installed by python development package.
```
```suggestion
Pip requires some header files and static libraries. Install the Python development package to get them:
```

iriark01 commented on docs/tools/mbed_cli_2/install.md:
````diff
@@ -0,0 +1,53 @@
+    python -m pip install mbed-tools --pre
+    ```
+
+<span class="tips">**Tip:** On a Unix-like system, you may need to ensure that `~/.local/bin` is accessible and in your path. </span>
````
```suggestion
<span class="tips">**Tip:** On a Unix-like system, ensure that `~/.local/bin` is accessible and in your path. </span>
```

jainvikas8 committed:
Update docs/tools/mbed_cli_2/install.md

Co-authored-by: Irit Arkin <irit.arkin@arm.com>

jainvikas8 committed:
Update docs/tools/mbed_cli_2/install.md

Co-authored-by: Irit Arkin <irit.arkin@arm.com>

Patater commented:
`feature-cmake` has been merged to mbed-os. We should merge this PR so that the new feature has docs.

jainvikas8 commented:
Forced pushed to resolve conflicts and rebase

Patater commented on docs/tools/mbed_cli_2/intro.md:
```diff
@@ -0,0 +1,7 @@
+# Mbed CLI 2
+
+Starting with version 6.x, Mbed OS is moving to Mbed CLI 2. It uses Ninja as a build system, and CMake to generate the build environment and manage the build process in a compiler-independent manner.
```
I believe the x should be 5. The first Mbed OS release with CMake support is Mbed OS 6.5.

```diff
- 6.x
+ 6.5
```

Patater commented on docs/tools/mbed_cli_2/intro.md:
```diff
@@ -0,0 +1,7 @@
+
+Mbed CLI 2 parses the Mbed OS build configuration and outputs it to a format CMake can read. It also provides a user friendly interface to CMake and Ninja so you can configure, generate and execute builds with a single command.
+
+<span class="notes">**Note:** You'll still need Mbed CLI 1 for older versions of Mbed OS (6.x and older). You can install both tools side by side.</span>
```
```diff
- 6.x
+ 6.4
```

Patater commented on docs/tools/tools_intro.md:
```diff
@@ -19,9 +19,28 @@ The Mbed Online Compiler is our in-house IDE and should be familiar to anyone wh
+Mbed OS has two versions of its command line tool:
+
+- For 6.y and newer, use Mbed CLI 1 or Mbed CLI 2.
+- For 6.x and older, use Mbed CLI 1.
```
```diff
- 6.x
+ 6.4
```

Patater commented on docs/tools/tools_intro.md:
```diff
@@ -19,9 +19,28 @@ The Mbed Online Compiler is our in-house IDE and should be familiar to anyone wh
-We created the Mbed command-line tool (Mbed CLI), a Python-based tool, specifically for Mbed OS. For more information, see the [Mbed CLI page](../tools/developing-mbed-cli.html).
+Mbed OS has two versions of its command line tool:
+
+- For 6.y and newer, use Mbed CLI 1 or Mbed CLI 2.
```
```diff
- 6.y
+ 6.5
```

Patater commented on docs/tools/tools_intro.md:
```diff
@@ -19,9 +19,28 @@ The Mbed Online Compiler is our in-house IDE and should be familiar to anyone wh
+
+For more information, see the [Mbed CLI 2 page](../build-tools/mbed-cli-2.html).
+
+<span class="notes">**Note:** You'll still need Mbed CLI 1 for older versions of Mbed OS (6.x and older). You can install both tools side by side.</span>
```
```diff
- 6.x
+ 6.4
```

Patater commented on docs/tools/tools_intro.md:
```diff
@@ -19,9 +19,28 @@ The Mbed Online Compiler is our in-house IDE and should be familiar to anyone wh
+
+<span class="notes">**Note:** You'll still need Mbed CLI 1 for older versions of Mbed OS (6.x and older). You can install both tools side by side.</span>
+
+### Mbed CLI 1 - Mbed OS 6.x and older
```
```diff
- 6.x
+ 6.4
```

Patater commented on docs/tools/tools_intro.md:
```diff
@@ -19,9 +19,28 @@ The Mbed Online Compiler is our in-house IDE and should be familiar to anyone wh
+
+You can have both tools installed side by side, so you can work with multiple versions of Mbed OS.
+
+### Mbed CLI 2 - Mbed OS 6.y and newer
```
```diff
- 6.y
+ 6.5
```

Patater commented on docs/tools/tools_intro.md:
```diff
@@ -19,9 +19,28 @@ The Mbed Online Compiler is our in-house IDE and should be familiar to anyone wh
+
+### Mbed CLI 2 - Mbed OS 6.y and newer
+
+Starting with version 6.x, Mbed OS is moving to Mbed CLI 2. It uses Ninja as a build system, and CMake to generate the build environment and manage the build process in a compiler-independent manner.
```
```diff
- 6.x
+ 6.5
```

jainvikas8 committed:
docs: Subsitute the right Mbed OS version

* Replace 6.x to 6.4 (Old version which doesn't support Mbed CLI 2)
* Replace 6.y to 6.5 (Supports Mbed CLI 2 and Mbed CLI 1)

Patater commented on docs/tools/tools_intro.md:
```diff
@@ -23,22 +23,22 @@ For more information, please see the [Online Compiler page](../build-tools/mbed-
+### Mbed CLI 2 - Mbed OS 6.5 and newer
 
-Starting with version 6.x, Mbed OS is moving to Mbed CLI 2. It uses Ninja as a build system, and CMake to generate the build environment and manage the build process in a compiler-independent manner.
+Starting with version 6.4, Mbed OS is moving to Mbed CLI 2. It uses Ninja as a build system, and CMake to generate the build environment and manage the build process in a compiler-independent manner.
```
```diff
- 6.4
+ 6.5
```
